monai/_extensions/hausdorff/hausdorff_cuda.cu

```diff
@@ -0,0 +1,2521 @@
+
+  ForBoolKernelArgs<T> fbArgs = executeHausdoff(
+      forFullBoolPrepArgs, WIDTH, HEIGHT, DEPTH, occData, stream1, false, robustnessPercent, resIterNeeded);
+
+  size_t sizeMinMax = sizeof(unsigned int) * 20;
+  // making sure we have all resultsto copy on cpu
+  cudaDeviceSynchronize();
+  cudaMemcpy(metaData.minMaxes, fbArgs.metaData.minMaxes, sizeMinMax, cudaMemcpyDeviceToHost);
+
+  int result = metaData.minMaxes[13];
+
+  cudaFreeAsync(fbArgs.minMaxes, stream1);
+  // free(minMaxesCPU);
+
+  cudaStreamDestroy(stream1);
+  cudaDeviceSynchronize();
+
+  return result;
+}
+
+template <typename T>
+at::Tensor getHausdorffDistance_CUDA_FullResList_local(
+    at::Tensor goldStandard,
+    at::Tensor algoOutput,
+    int WIDTH,
+    int HEIGHT,
+    int DEPTH,
+    float robustnessPercent,
+    at::Tensor compare_values) {
+  // TODO() use https ://pytorch.org/cppdocs/notes/tensor_cuda_stream.html
+  cudaStream_t stream1;
+  cudaStreamCreate(&stream1);
+
+  MetaDataCPU metaData;
+  // size_t size = sizeof(unsigned int) * 20;
+  // unsigned int* minMaxesCPU = (unsigned int*)malloc(size);
+  // metaData.minMaxes = minMaxesCPU;
+
+  ForFullBoolPrepArgs<T> forFullBoolPrepArgs;
+  forFullBoolPrepArgs.metaData = metaData;
+  forFullBoolPrepArgs.compare_values = compare_values.item<T>();
+  forFullBoolPrepArgs.goldArr = goldStandard;
+  forFullBoolPrepArgs.segmArr = algoOutput;
+
+  occupancyCalcData occData = getOccupancy<T>();
+
+  ForBoolKernelArgs<T> fbArgs =
+      executeHausdoff(forFullBoolPrepArgs, WIDTH, HEIGHT, DEPTH, occData, stream1, false, robustnessPercent, true);
+
+  cudaFreeAsync(fbArgs.minMaxes, stream1);
+  // free(metaData.minMaxesCPU);
+
+  cudaStreamDestroy(stream1);
+
+  return fbArgs.resultListPointerIterNumbTensor;
+}
+
+/*
+Functions for pybind
+*/
+int getHausdorffDistance_CUDA(
+    at::Tensor goldStandard,
+    at::Tensor algoOutput,
+    const int WIDTH,
+    const int HEIGHT,
+    const int DEPTH,
+    const float robustnessPercent,
+    at::Tensor compare_values) {
+  int res = 0;
+
+  AT_DISPATCH_ALL_TYPESWithBool(
+      goldStandard.type(), "getHausdorffDistance_CUDA", ([&] {
+        res = getHausdorffDistance_CUDA_Generic<scalar_t>(
+            goldStandard, algoOutput, WIDTH, HEIGHT, DEPTH, robustnessPercent, false, compare_values);
+      }));
+  return res;
+}
+
+at::Tensor getHausdorffDistance_CUDA_FullResList(
+    at::Tensor goldStandard,
+    at::Tensor algoOutput,
+    const int WIDTH,
+    const int HEIGHT,
+    const int DEPTH,
+    const float robustnessPercent,
+    at::Tensor compare_values) {
+  at::Tensor res;
+  AT_DISPATCH_ALL_TYPESWithBool(
+      goldStandard.type(), "getHausdorffDistance_CUDA_FullResList", ([&] {
+        res = getHausdorffDistance_CUDA_FullResList_local<scalar_t>(
+            goldStandard, algoOutput, WIDTH, HEIGHT, DEPTH, robustnessPercent, compare_values);
+      }));
+
+  return res;
+}
+
+
+
+
+
+
+
+<<<<<<< HEAD
+=======
+
```

Comment:
there are some compilation errors such as this one from git diff.... https://github.com/Project-MONAI/MONAI/runs/5971824137?check_suite_focus=true

```
======================================================================
ERROR: test_value_00 (tests.test_hausdorff_distance_morphologic.TestHausdorffDistanceMorphological)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/torch/utils/cpp_extension.py", line 1539, in _run_ninja_build
    env=env)
  File "/usr/lib/python3.7/subprocess.py", line 512, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['ninja', '-v']' returned non-zero exit status 1.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/parameterized/parameterized.py", line 533, in standalone_func
    return func(*(a + p.args), **p.kwargs)
  File "/__w/MONAI/MONAI/tests/test_hausdorff_distance_morphologic.py", line 119, in test_value
    compare_values.to(device), percentt, True
  File "/__w/MONAI/MONAI/monai/metrics/morphological_hausdorff_distance.py", line 90, in __init__
    self.compiled_extension = load_module("hausdorff")
  File "/__w/MONAI/MONAI/monai/_extensions/loader.py", line 87, in load_module
    name=name, sources=source, extra_cflags=define_args, extra_cuda_cflags=define_args, verbose=verbose_build
  File "/usr/local/lib/python3.7/dist-packages/torch/utils/cpp_extension.py", line 997, in load
    keep_intermediates=keep_intermediates)
  File "/usr/local/lib/python3.7/dist-packages/torch/utils/cpp_extension.py", line 1202, in _jit_compile
    with_cuda=with_cuda)
  File "/usr/local/lib/python3.7/dist-packages/torch/utils/cpp_extension.py", line 1300, in _write_ninja_file_and_build_library
    error_prefix="Error building extension '{}'".format(name))
  File "/usr/local/lib/python3.7/dist-packages/torch/utils/cpp_extension.py", line 1555, in _run_ninja_build
    raise RuntimeError(message) from e
RuntimeError: Error building extension 'hausdorff_Linux_3_7_5_17_10_2': [1/2] /usr/local/cuda/bin/nvcc -DTORCH_EXTENSION_NAME=hausdorff_Linux_3_7_5_17_10_2 -DTORCH_API_INCLUDE_EXTENSION_H -DPYBIND11_COMPILER_TYPE=\"_gcc\" -DPYBIND11_STDLIB=\"_libstdcpp\" -DPYBIND11_BUILD_ABI=\"_cxxabi1011\" -isystem /usr/local/lib/python3.7/dist-packages/torch/include -isystem /usr/local/lib/python3.7/dist-packages/torch/include/torch/csrc/api/include -isystem /usr/local/lib/python3.7/dist-packages/torch/include/TH -isystem /usr/local/lib/python3.7/dist-packages/torch/include/THC -isystem /usr/local/cuda/include -isystem /usr/include/python3.7m -D_GLIBCXX_USE_CXX11_ABI=0 -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS__ --expt-relaxed-constexpr -gencode=arch=compute_70,code=sm_70 --compiler-options '-fPIC' -std=c++14 -c /__w/MONAI/MONAI/monai/_extensions/hausdorff/hausdorff_cuda.cu -o hausdorff_cuda.cuda.o 
FAILED: hausdorff_cuda.cuda.o 
/usr/local/cuda/bin/nvcc -DTORCH_EXTENSION_NAME=hausdorff_Linux_3_7_5_17_10_2 -DTORCH_API_INCLUDE_EXTENSION_H -DPYBIND11_COMPILER_TYPE=\"_gcc\" -DPYBIND11_STDLIB=\"_libstdcpp\" -DPYBIND11_BUILD_ABI=\"_cxxabi1011\" -isystem /usr/local/lib/python3.7/dist-packages/torch/include -isystem /usr/local/lib/python3.7/dist-packages/torch/include/torch/csrc/api/include -isystem /usr/local/lib/python3.7/dist-packages/torch/include/TH -isystem /usr/local/lib/python3.7/dist-packages/torch/include/THC -isystem /usr/local/cuda/include -isystem /usr/include/python3.7m -D_GLIBCXX_USE_CXX11_ABI=0 -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS__ --expt-relaxed-constexpr -gencode=arch=compute_70,code=sm_70 --compiler-options '-fPIC' -std=c++14 -c /__w/MONAI/MONAI/monai/_extensions/hausdorff/hausdorff_cuda.cu -o hausdorff_cuda.cuda.o 
/__w/MONAI/MONAI/monai/_extensions/hausdorff/hausdorff_cuda.cu:22:10: fatal error: cuda/pipeline: No such file or directory
 #include <cuda/pipeline>
          ^~~~~~~~~~~~~~~
compilation terminated.
ninja: build stopped: subcommand failed.
```

Reply:
For what version of CUDA it is compiled? pipelines are relatively new feature

Reply:
hmm I must see wheather is there a beckward compatibility pipeline - Nvidia frequently provides sth like that - in the meantime I do not understand wwhat does the linter complain about

Reply:
you have a bash terminal you can run the linter locally such as `./runtests.sh -f`. and unit tests with `./runtests.sh -u`. `./runtests.sh --help` gives you more options.

Reply:
O I am looking for backward compatible ways, Hovewer is it possible to flag it as available only for CUDA 11 ?

Reply:
good question, currently that mechanism is not implemented. I'd suggest we add a check in `test_hausdorff_distance_morphologic.py` for the cuda version, like this one:
https://github.com/Project-MONAI/MONAI/blob/42159ffd3a3003efdee807683922ea1cec4cb27a/tests/utils.py#L148
we skip the test if the version condition not satisfied.

Reply:
thanks - done